File: backend/app/models/session.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class SessionType(str, enum.Enum):
    group = "group"
    solo = "solo"
    outreach = "outreach"


class ConductedBy(str, enum.Enum):
    asha_session = "asha_session"
    independent = "independent"


class SessionStatus(str, enum.Enum):
    scheduled = "scheduled"
    in_progress = "in_progress"
    completed = "completed"
    missed = "missed"


class ResponseMarking(str, enum.Enum):
    independent = "independent"
    hint = "hint"
    no_response = "no_response"


class Session(Base):
    """A session is not implicitly one patient — group sessions carry multiple patient_ids with per-patient response tracking."""

    __tablename__ = "sessions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    type: Mapped[SessionType] = mapped_column(Enum(SessionType), nullable=False)
    conducted_by: Mapped[ConductedBy] = mapped_column(Enum(ConductedBy), nullable=False)
    asha_id: Mapped[str | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    scheduled_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[SessionStatus] = mapped_column(Enum(SessionStatus), default=SessionStatus.scheduled)
    # patient_ids is stored as a comma-separated string here for SQLite/Postgres portability;
    # a join table is the natural upgrade once the sync service is wired up.
    patient_ids: Mapped[str] = mapped_column(String, nullable=False)


class ResponseRecord(Base):
    """Core telemetry: one row per patient, per round, per session. Feeds the adaptive engine and baseline comparison."""

    __tablename__ = "response_records"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id: Mapped[str] = mapped_column(ForeignKey("sessions.id"), nullable=False)
    patient_id: Mapped[str] = mapped_column(ForeignKey("patients.id"), nullable=False)
    round_number: Mapped[int] = mapped_column(Integer, nullable=False)
    marking: Mapped[ResponseMarking] = mapped_column(Enum(ResponseMarking), nullable=False)
    game_module: Mapped[str] = mapped_column(String, nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

File: frontend/lib/core/models/session.dart
enum SessionType { group, solo, outreach }

enum ConductedBy { ashaSession, independent }

enum SessionStatus { scheduled, inProgress, completed, missed }

/// The three-way marking an ASHA logs for each patient, each round, during a session.
enum ResponseMarking { independent, hint, noResponse }

String sessionTypeToJson(SessionType t) => t.name;

String conductedByToJson(ConductedBy c) => switch (c) {
      ConductedBy.ashaSession => 'asha_session',
      ConductedBy.independent => 'independent',
    };

String responseMarkingToJson(ResponseMarking m) => switch (m) {
      ResponseMarking.independent => 'independent',
      ResponseMarking.hint => 'hint',
      ResponseMarking.noResponse => 'no_response',
    };

/// SessionStatus.values.byName() would throw on the backend's snake_case
/// "in_progress" (Dart's enum name is camelCase inProgress) — map explicitly.
SessionStatus sessionStatusFromJson(String value) => switch (value) {
      'scheduled' => SessionStatus.scheduled,
      'in_progress' => SessionStatus.inProgress,
      'completed' => SessionStatus.completed,
      'missed' => SessionStatus.missed,
      _ => throw ArgumentError('Unknown session status: $value'),
    };

String sessionStatusToJson(SessionStatus s) => switch (s) {
      SessionStatus.scheduled => 'scheduled',
      SessionStatus.inProgress => 'in_progress',
      SessionStatus.completed => 'completed',
      SessionStatus.missed => 'missed',
    };

ResponseMarking responseMarkingFromJson(String value) => switch (value) {
      'independent' => ResponseMarking.independent,
      'hint' => ResponseMarking.hint,
      'no_response' => ResponseMarking.noResponse,
      _ => throw ArgumentError('Unknown response marking: $value'),
    };

class CareSession {
  final String id;
  final SessionType type;
  final ConductedBy conductedBy;
  final String? ashaId;
  final DateTime scheduledTime;
  final SessionStatus status;
  final List<String> patientIds;

  const CareSession({
    required this.id,
    required this.type,
    required this.conductedBy,
    required this.scheduledTime,
    required this.status,
    required this.patientIds,
    this.ashaId,
  });

  bool get isGroup => type == SessionType.group;

  CareSession copyWith({SessionStatus? status}) => CareSession(
        id: id,
        type: type,
        conductedBy: conductedBy,
        ashaId: ashaId,
        scheduledTime: scheduledTime,
        status: status ?? this.status,
        patientIds: patientIds,
      );

  factory CareSession.fromJson(Map<String, dynamic> json) => CareSession(
        id: json['id'] as String,
        type: SessionType.values.byName(json['type'] as String),
        conductedBy: (json['conducted_by'] as String) == 'asha_session'
            ? ConductedBy.ashaSession
            : ConductedBy.independent,
        ashaId: json['asha_id'] as String?,
        scheduledTime: DateTime.parse(json['scheduled_time'] as String),
        status: sessionStatusFromJson(json['status'] as String),
        patientIds: List<String>.from(json['patient_ids'] as List),
      );

  Map<String, dynamic> toLocalJson() => {
        'id': id,
        'type': sessionTypeToJson(type),
        'conducted_by': conductedByToJson(conductedBy),
        'asha_id': ashaId,
        'scheduled_time': scheduledTime.toIso8601String(),
        'status': sessionStatusToJson(status),
        'patient_ids': patientIds,
      };

  /// No local-only fields yet, so this is just an alias — kept distinct from
  /// fromJson() so callers name their intent, matching toLocalJson() above.
  factory CareSession.fromLocalJson(Map<String, dynamic> json) => CareSession.fromJson(json);
}

class ResponseRecord {
  final String sessionId;
  final String patientId;
  final int roundNumber;
  final ResponseMarking marking;
  final String gameModule;
  final DateTime timestamp;

  const ResponseRecord({
    required this.sessionId,
    required this.patientId,
    required this.roundNumber,
    required this.marking,
    required this.gameModule,
    required this.timestamp,
  });

  Map<String, dynamic> toJson() => {
        'session_id': sessionId,
        'patient_id': patientId,
        'round_number': roundNumber,
        'marking': responseMarkingToJson(marking),
        'game_module': gameModule,
        'timestamp': timestamp.toIso8601String(),
      };

  factory ResponseRecord.fromJson(Map<String, dynamic> json) => ResponseRecord(
        sessionId: json['session_id'] as String,
        patientId: json['patient_id'] as String,
        roundNumber: json['round_number'] as int,
        marking: responseMarkingFromJson(json['marking'] as String),
        gameModule: json['game_module'] as String,
        timestamp: DateTime.parse(json['timestamp'] as String),
      );
}
